### PR TITLE
Simple test run improvements

### PR DESCRIPTION
**Pull request by salieri11**
This PR adresses #16 and adds ability to run Simple Test via Python script, which supports various BFT configurations.
Please refer to [bftengine/tests/simpleTest/README.md](bftengine/tests/simpleTest/README.md) for details.

### CODE REVIEW

File: bftengine/tests/simpleTest/scripts/simpleTest.py [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+                       "c - max. number of slow replicas, " +
+                       "cl - number of running clients, " +
+                       "testViewChange - if present, viewChange will be " +
+                       "triggered after 1/2 of iterations done. " +
+                       "Please note, if View change is enabled AND n - r " +
+                       "<= f - the system will not progress after primary " +
+                       "is killed, and the test will SUCCESS after timeout " +
+                       "- this is desired behaviour for this configuration." +
+                       "You can provide set " +
+                       "set of configurations in the form of " +
```
Comment (upshaw-alex) [inline bftengine/tests/simpleTest/scripts/simpleTest.py]: nit: The word "set" is repeated here.

Furthermore, I think it would read a little better as "**a** set" rather than "set".

File: bftengine/tests/simpleTest/client.cpp [hunk 2]
```diff
@@ -103,7 +151,14 @@ int main(int argc, char **argv) {
   // operation.
   bool hasExpectedLastValue = false;
 
-  for (int i = 1; i <= numOfOperations; i++) {
+  LOG_INFO(clientLogger, "Starting " << cp.numOfOperations);
+
+  for (int i = 1; i <= cp.numOfOperations; i++) {
+    if(i > 0 && i % 100 == 0) {
+      printf("Iterations count: 100\n");
+      printf("Total iterations count: %i\n", i);
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/tests/simpleTest/client.cpp]: These are special outputs, I seek for them in the python code, that's why I use plain printf. Wil add comment to explain.

File: bftengine/tests/simpleTest/client.cpp [hunk 3]
```diff
@@ -62,12 +109,12 @@ int main(int argc, char **argv) {
   config.configure();
 #endif
 
+  ClientParams cp = parse_params(argc, argv);
+  LOG_INFO(clientLogger, "ClientParams: clientId: " << cp.clientId << ", numOfReplicas: " << cp.numOfReplicas << ", numOfClients: " << cp.numOfClients << ", numOfIterations: " << cp.numOfOperations << ", fVal: " << cp.numOfFaulty << ", cVal: " << cp.numOfSlow);
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/tests/simpleTest/client.cpp]: done

File: bftengine/tests/simpleTest/client.cpp [hunk 4]
```diff
@@ -50,8 +53,52 @@ using bftEngine::SeqNumberGeneratorForClientRequests;
...
+      cp.clientId = std::stoi(argv[i + 1]);
+    else if(p == "-r")
+      cp.numOfReplicas = std::stoi(argv[i + 1]);
+    else if(p == "-cl")
+      cp.numOfClients = std::stoi(argv[i + 1]);
+    else if(p == "-c")
+      cp.numOfSlow = std::stoi(argv[i + 1]);
+    else if(p == "-f")
+      cp.numOfFaulty = std::stoi(argv[i + 1]);
+
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/tests/simpleTest/client.cpp]: added validations

File: bftengine/src/communication/PlainUDPCommunication.cpp [hunk 5]
```diff
@@ -318,11 +327,12 @@ class PlainUDPCommunication::PlainUdpImpl {
     if (res == addr2nodes.end()) {
       // IG: if we don't know the sender we just ignore this message and
       // continue.
-      LOG_WARN(_logger, "Unknown sender, address: " << key);
-      return NodeAddressResolveResult({0, false});
+      LOG_ERROR(_logger, "Unknown sender, address: " << key);
+      return NodeAddressResolveResult({0, false,key});
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/src/communication/PlainUDPCommunication.cpp]: done

File: bftengine/tests/simpleTest/replica.cpp [hunk 6]
```diff
@@ -74,15 +101,95 @@ using bftEngine::PlainTcpConfig;
...
+        i++;
+      }
+      else if (p == "-vc") {
+        rp.viewChangeEnabled = true;
+        i++;
+      }
+      else if (p == "-vct") {
+        rp.viewChangeTimeout = std::stoi(argv[i + 1]);
+        i+=2;
+      }
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/tests/simpleTest/replica.cpp]: added validations

File: bftengine/src/communication/PlainTcpCommunication.cpp [hunk 7]
```diff
@@ -931,11 +926,14 @@ class PlainTCPCommunication::PlainTcpImpl {
 
   virtual ~PlainTcpImpl() {
     LOG_TRACE(_logger, "PlainTCPDtor");
+    _pIoThread = nullptr;
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/src/communication/PlainTcpCommunication.cpp]: its done only for being very explicit on the destruction. you are right, it doesn't add anything to the destruction. I will remove it to make it less confusing - but just in case - I always prefer to do things explicitly, even if they are meaningless somehow. It gives better semantics sometimes.

File: bftengine/src/communication/PlainTcpCommunication.cpp [hunk 8]
```diff
@@ -931,11 +926,14 @@ class PlainTCPCommunication::PlainTcpImpl {
...
   }
 };
 
 PlainTCPCommunication::~PlainTCPCommunication() {
-  _ptrImpl->Stop();
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/src/communication/PlainTcpCommunication.cpp]: It should be stopped using Stop() method. Dtor only cleans up

File: bftengine/tests/simpleTest/scripts/simpleTest.py [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+
+    def invoke_on_new_line(self, txt):
+        if self.on_new_line_callback:
+            self.on_new_line_callback=self.on_new_line_callback(txt, self.uid)
+
+    def collect(self):
+        with io.open(self.out_file, "r") as reader:
+            log = reader.read()
+
+        m = re.search("(?<=\n).*(?=\n$)", log)
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/tests/simpleTest/scripts/simpleTest.py]: m stands for Match in regex related operations, AFAIK exactly the same as i for loops

File: bftengine/tests/simpleTest/client.cpp [hunk 10]
```diff
@@ -50,8 +53,52 @@ using bftEngine::SeqNumberGeneratorForClientRequests;
...
+  int numOfFaulty = 1;
+  int numOfSlow = 0;
+};
+
+ClientParams parse_params(int argc, char** argv) {
+  ClientParams cp;
+  if(argc < 2)
+    return cp;
+
+  for(int i=1; i < argc;) {
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/tests/simpleTest/client.cpp]: i added the validations on the values' overflow. these binaries are not expected to be executed manually so I don't take care of the edge cases.

File: bftengine/tests/simpleTest/scripts/simpleTest.py [hunk 11]
```diff
@@ -0,0 +1,465 @@
...
+    @staticmethod
+    def reset_globals():
+        TestConfig.iterations_done = 0
+        TestConfig.leader_killed = False
+        TestConfig.config_run_timer = None
+        TestConfig.config_last_update_time = None
+
+    def prepare(self):
+        exec_dir = os.path.abspath(self.env.binary_dir)
+        time_stamp = time.strftime("%d_%m_%Y_%H_%M_%S")
```
Comment (beerriot) [inline bftengine/tests/simpleTest/scripts/simpleTest.py]: Please change this to an ISO8601-like format, so that the directories sort lexicographically. e.g. `%Y%m%dT%H%M%S`

File: bftengine/tests/simpleTest/scripts/simpleTest.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply bftengine/tests/simpleTest/scripts/simpleTest.py]: Okay, thank you for informing me; sorry I was not already aware of that.

File: bftengine/tests/simpleTest/scripts/simpleTest.py [hunk 12]
```diff
@@ -0,0 +1,465 @@
...
+
+    # this method is used to determine total number of iterations been proceed
+    # so far by clients and to kill primary when needed
+    def on_new_line(self, text, instanceId):
+        if text is None:
+            return
+        pattern = re.compile(r"(?<=^Iterations count:\s)\d+")
+        TestConfig.iterations_counter_lock.acquire()
+        found = False
+        for m in re.finditer(pattern, text):
```
[the remark this answers is not in the document]
Comment (upshaw-alex) [reply bftengine/tests/simpleTest/scripts/simpleTest.py]: This is no longer an issue; see above discussion on line 94.

**Comment by salieri11** [comment]
> `./simpleTest.py`: `TEST SUCCESS` after 9.440s.
good

> `./simpleTest.py -i 5000 -nc -bft n=6,r=6,f=1,c=1,cl=2,testViewChange`: `TEST SUCCESS` after 2m9.670s. This test has a significant delay after 5000 iterations when it kills replica 0.
good. the View Change is triggered after 60 seconds, by default (the -vct parameter)

> `./simpleTest.py -i 2000 -nc -bft n=20,r=15,f=5,c=2,cl=4`: `TEST SUCCESS` after 1m22.376s.
good

> `./simpleTest.py -i 2000 -nc -bft n=20,r=13,f=5,c=2,cl=4,testViewChange`: `TEST SUCCESS` after 6m53.064s. This test has a substantial delay after 4000 iterations when it kills replica 0.
very good. in this configuration view change should NOT complete since 13 is the minimal number of replicas for the system to work (out of 20 replicas) and when primary is killed - it will never complete. The test timed out after (60 sec * 5) + 1 seconds = 4 - this is general timeout for the test, and in this case this is EXPECTED behavior.

> `./simpleTest.py -l DEBUG -i 5000 -nc -bft n=6,r=6,f=1,c=1,cl=2,testViewChange`: `TEST SUCCESS` after 2m3.084s. This test has a period of not receiving any data from clients after 5000 iterations when it kills replica 0.
good, same as 2

> `./simpleTest.py -l DEBUG -i 2000 -nc -bft n=20,r=15,f=5,c=2,cl=4`: `TEST SUCCESS` after 1m20.289s.
good

> `./simpleTest.py -l DEBUG -i 2000 -nc -bft n=20,r=13,f=5,c=2,cl=4,testViewChange`: `TEST SUCCESS` after 7m3.348s. This test has a long period of not receiving data from clients after 4000 iterations when it kills replica 0.
good, same as 4

File: bftengine/tests/simpleTest/client.cpp [hunk 13]
```diff
@@ -50,8 +54,106 @@ using bftEngine::SeqNumberGeneratorForClientRequests;
...
+  } catch (std::invalid_argument &e) {
+    printf("Parameters should be integers only\n");
+    exit(-1);
+  } catch (std::out_of_range &e) {
+    printf("One of the parameters is out of range\n");
+    exit(-1);
+  }
+
+  if(3 * cp.numOfFaulty + 2 * cp.numOfSlow + 1 != cp.numOfReplicas) {
+    printf("Number of replicas is not 3f + 2c + 1");
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/tests/simpleTest/client.cpp]: 10x

**Comment by beerriot** [comment]
I just tried a fresh _debug_ build, and saw the same stuck behavior. However, I'm now trying a fresh _release_ build, and the test is progressing as expected. If either @salieri11 or @upshaw-alex is seeing the test progress on a _debug_ build, then it must just be a configuration issue on my system, and I'll be happy to +1 for merge.

**Comment by salieri11** [comment]
> I just tried a fresh _debug_ build, and saw the same stuck behavior. However, I'm now trying a fresh _release_ build, and the test is progressing as expected. If either @salieri11 or @upshaw-alex is seeing the test progress on a _debug_ build, then it must just be a configuration issue on my system, and I'll be happy to +1 for merge.

I work in debug most of the time, checking Release only when issuing PR - and in both cases it runs. @upshaw-alex , can you please check?

File: bftengine/tests/simpleTest/scripts/simpleTest.py [hunk 14]
```diff
@@ -457,6 +458,15 @@ def main():
         if n is None or r is None or f is None or c is None or cl is None:
             print("Unable to parse BFT parameters")
             return -1
+        if 3 * f + 2 * c + 1 != n:
+            g_logger.error("N = 3f + 2c + 1 is not satisfied")
+            return -1
+        max_r = max(4, multiprocessing.cpu_count() * 4)
+        if r > max_r:
+            g_logger.error("max number of running replicas on your machine is "
```
Comment (upshaw-alex) [inline bftengine/tests/simpleTest/scripts/simpleTest.py]: Is the limit of 2 running replicas per core just to keep this test from being too slow, or is this limit required for correctness?
